main.cpp

```diff
@@ -179,129 +122,268 @@ int main()
+    DoubleType& divide(const DoubleType& dt);
+    DoubleType& add(const IntType& it);
+    DoubleType& subtract(const IntType& it);
+    DoubleType& multiply(const IntType& it);
+    DoubleType& divide(const IntType& it);
 };
 
-double DoubleType::add( double lhs, double rhs )
+DoubleType& DoubleType::add( double rhs )
 {
-    return lhs + rhs;
+    *value += rhs;
+    return *this;
 }
-double DoubleType::subtract( double lhs, double rhs )
+DoubleType& DoubleType::subtract( double rhs )
 {
-    return lhs - rhs;
+    *value -= rhs;
+    return *this;
 }
-double DoubleType::multiply( double lhs, double rhs )
+DoubleType& DoubleType::multiply( double rhs )
 {
-    return lhs * rhs;
+    *value *= rhs;
+    return *this;
 }
-double DoubleType::divide( double lhs, double rhs )
+DoubleType& DoubleType::divide( double rhs )
 {
     if (rhs == 0.0)
-        std::cout << "\n" << "warning, floating point division by zero returns 'inf' !" << "\n";
+        std::cout << "warning: floating point division by zero!" << "\n";
         
-    return lhs / rhs;
+    *value /= rhs;
+    return *this;
 }
 
 struct IntType
 {
-    int add( int lhs, int rhs );
-    int subtract( int lhs, int rhs );
-    int multiply( int lhs, int rhs );
-    int divide( int lhs, int rhs );
+    IntType(int it) : value(new int(it)) {}
+    ~IntType()
+    {
+        delete value;
+        value = nullptr;
+    }
+
+    int* value;
+
+    IntType& add( int rhs );
+    IntType& subtract( int rhs );
+    IntType& multiply( int rhs );
+    IntType& divide( int rhs );
+
+    IntType& add(const FloatType& ft);
+    IntType& subtract(const FloatType& ft);
+    IntType& multiply(const FloatType& ft);
+    IntType& divide(const FloatType& ft);
+    IntType& add(const DoubleType& dt);
+    IntType& subtract(const DoubleType& dt);
+    IntType& multiply(const DoubleType& dt);
+    IntType& divide(const DoubleType& dt);
+    IntType& add(const IntType& it);
+    IntType& subtract(const IntType& it);
+    IntType& multiply(const IntType& it);
+    IntType& divide(const IntType& it);
 };
 
-int IntType::add( int lhs, int rhs )
+IntType& IntType::add( int rhs )
 {
-    return lhs + rhs;
+    *value += rhs;
+    return *this;
 }
-int IntType::subtract( int lhs, int rhs )
+IntType& IntType::subtract( int rhs )
 {
-    return lhs - rhs;
+    *value -= rhs;
+    return *this;
 }
-int IntType::multiply( int lhs, int rhs )
+IntType& IntType::multiply( int rhs )
 {
-    return lhs * rhs;
+    *value *= rhs;
+    return *this;
 }
-int IntType::divide( int lhs, int rhs )
+IntType& IntType::divide( int rhs )
 {
     if (rhs == 0)
     {
-        std::cout << "error, integer division by zero will crash the program!" << "\n" << "returning lhs" << "\n";
-        return lhs;
+        std::cout << "error: integer division by zero is an error and will crash the program!\n";
+        return *this;
     }
 
-    return lhs / rhs;
+    *value /= rhs;
+    return *this;
 }
```

Comment:
You have two `return *this;` statements in this function
Try to revise the logic so it is only written once.
- [ ] revise